### PR TITLE
Initial documentation

### PR DESCRIPTION
**Pull request by pyhys**

### CODE REVIEW

File: include/pnet_api.h [hunk 1]
```diff
@@ -465,20 +465,6 @@ typedef struct pnet_result
    uint16_t                add_data_2;    /**< Application-specific error information. */
 } pnet_result_t;
 
-/**
```
Comment (hefloryd) [inline include/pnet_api.h]: Is this information (header files) incorrect or obsoleted by the documentation above?

File: README.md [hunk 2]
```diff
@@ -1,15 +1,26 @@
-p-net: PROFINET device stack
-============================
+p-net Profinet stack
```
Comment (hefloryd) [inline README.md]: The stack is only for Profinet devices, should we make that clear here?

File: README.md [hunk 3]
```diff
@@ -30,97 +41,37 @@ Features:
...
-The clang static analyzer can also be used if installed. From a clean
-build directory, run:
 
-```console
-user@host:~/build$ scan-build cmake $repo
-user@host:~/build$ scan-build make
-```
+Requirements
+------------
+The the platform must be able to send and receive raw Ethernet Layer 2 frames,
```
Comment (hefloryd) [inline README.md]: The the

File: README.md [hunk 4]
```diff
@@ -30,97 +41,37 @@ Features:
...
+and the Ethernet driver must be able to handle frames of size ?? Bytes. It
```
Comment (hefloryd) [inline README.md]: Add a value for ??

File: README.md [hunk 5]
```diff
@@ -30,97 +41,37 @@ Features:
...
-=========
+* cmake 3.13 or later
 
- * Workbench 2017.1 or later
+For Linux:
 
-Set the following environment variables. You should use a bash shell,
-such as for instance the Command Line in your Toolbox installation.
+* gcc 4.6 or later
+* Take precautions to improve timings
```
Comment (hefloryd) [inline README.md]: This is a little vague (take precautions), can we refer to the chapter on linux timing?

File: doc/getting_started_rtkernel.rst [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+Create a build directory (on same level as p-net directory) and run cmake::
+
+    mkdir build
+    cd build
+    export COMPILERS=/opt/rt-tools/compilers
+    export RTK=<PATH TO YOUR MODIFIED>/rt-kernel-xmc4/
+    export BSP=xmc48relax
+
+TODO::
```
Comment (hefloryd) [inline doc/getting_started_rtkernel.rst]: Why is this a TODO?

File: include/pnet_api.h (hunk 1 above)
Comment (pyhys) [reply include/pnet_api.h]: This additional text field is confusing Doxygen (it is not connected to a function etc). The info is moved to to the freetext part of the documentation.

File: doc/capturing_packets.rst [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+Wireshark
+---------
+To install a relatively new Wireshark version on Ubuntu::
+
+    sudo add-apt-repository ppa:wireshark-dev/stable
+    sudo apt update
+    sudo apt -y install wireshark
+
+You must run it as root to be able to capture packets::
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/capturing_packets.rst]: Added comment on wireshark group, and link to instructions.

File: doc/compliancetest.rst [hunk 8]
```diff
@@ -0,0 +1,77 @@
+Compliance testing of IO-devices
+================================
+All Profinet products needs to be certified. That is done after testing by an
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/compliancetest.rst]: Done

File: doc/compliancetest.rst [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+tab. For the "Packet priority & VLAN" select "Packet priority & VLAN Disable".
+More details are given in the "Product Documentation" document for the tool.
+
+You might also need to turn off LLDP protocol for the selected network
+interface. Both Windows and Simatic TIA can have LLDP implemented.
+
+
+Set IP address of the Windows PC
+--------------------------------
+Which values?
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/compliancetest.rst]: Done

File: doc/creating_gsdml_files.rst [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+* ``MaxPortRxDelay="350"`` Time delay in ns needed for receiving frames
+* ``MaxPortTxDelay="160"`` Time delay in ns needed for sending frames
+
+Use an ``<MAUTypeItem>`` element to describe the Medium Attachment Unit type,
+which can be radio (0), copper at 100 Mbit/s (16) or fiber optics.
+
+
+Details on the module list
+--------------------------
+Profinet field devices can have different hardware modules, why there is a
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/creating_gsdml_files.rst]: Done

File: doc/creating_gsdml_files.rst [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+The element ``<ModuleInfo>`` has information on the module name in its
+subelements. The elements ``<HardwareRelease>`` and ``<SoftwareRelease>`` have
+``Value`` attributes.
+
+Each ``<SubmoduleItemRef>`` element has the attributes ``SubmoduleItemTarget``
+(which is a reference to a submodule) and ``AllowedInSubslots`` (which is a
+space separated list of subslot numbers).
+
+Virtual submodules are submodules that are built-in into a module (no physical
+submodule can be removed). Then is the ``PhysicalSubslots`` attribute not given
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/creating_gsdml_files.rst]: Modified

File: doc/creating_gsdml_files.rst [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+The subslot number for the first interface is 0x8000, and next interface (if
+any) has subslot number 0x8100. The first port of the first interface has
+subslot 0x8001, and next port of that interface has subslot number 0x8002.
+
+Interfaces are described using the ``<InterfaceSubmoduleItem>`` element, which
+has these attributes:
+
+* ``ID="IDS_I"``
+* ``SubmoduleIdentNumber="0x00000002"``
+* ``SubslotNumber="32768"`` This is 0x8000 = First interface
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/creating_gsdml_files.rst]: Commented which fields should be in hex and which should be decimal.

File: doc/general_profinet.rst [hunk 13]
```diff
@@ -0,0 +1,211 @@
+General on Profinet
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/general_profinet.rst]: Done

File: doc/general_profinet.rst [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+Net load class:
+
+* I
+* II
+* III
+
+
+Slots and modules
+-----------------
+Profinet IO-device has typically a number of slots where (hardware) modules can
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/general_profinet.rst]: Done

File: doc/getting_started_linux.rst [hunk 15]
```diff
@@ -0,0 +1,129 @@
+Getting started on Linux
+========================
+Running the IO-device application on Linux
+
+
+Install dependencies
+--------------------
+Install a recent version of cmake (at least 3.13)::
+
+    sudo snap install cmake --classic
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_linux.rst]: Added text on cmake in Ubuntu, and official repo.

File: doc/getting_started_linux.rst [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+When using it, you might need to replace ``cmake`` with ``snap run cmake``.
+If uninstalling the default ``cmake``, the snap-installed version should be
+used by default after reboot.
+
+Other dependencies::
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_linux.rst]: Commented that they are optional (for building documentation etc)

File: doc/getting_started_linux.rst [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+   scan-build cmake ..
+   scan-build make
+
+
+Run the Linux demo application
+------------------------------
+You need to set the IP-address (adapt network name and IP address)::
+
+   sudo ifconfig eth0 192.168.137.4 netmask 255.255.255.0 up
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_linux.rst]: Modified text ("If you need to use static IP ...")

File: doc/getting_started_linux.rst [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+        -s NAME      Set station name. Defaults to rt-labs-dev
+        -l FILE      Path to control LED. Defaults to not control any LED.
+        -b FILE      Path to read button1. Defaults to not read button1.
+        -d FILE      Path to read button2. Defaults to not read button2.
+
+Run the sample application::
+
+    cd build
+    make pn_dev
+    sudo ifconfig eth0 192.168.137.4 netmask 255.255.255.0 up
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_linux.rst]: Kept. Maybe remove?

File: doc/getting_started_rtkernel.rst [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+adapt paths etc accordingly.
+
+
+Install tools on a Linux Laptop
+-------------------------------
+Install (see description elsewhere):
+
+* Workbench (Version 2017.1 or later)
+* Compiler
+* RT-kernel, use variant ``rt-kernel-xmc4``
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Done

File: doc/getting_started_rtkernel.rst [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+Use the menu "File > Import". Select "General > Existing Projects". Click
+"Browse" and select the ``build`` directory that was created earlier.
+
+Use the menu "Project > Build All" to build it.
+
+
+More Workbench settings
+-----------------------
+If you intend to edit the p-net source code in the Workbench tool, you also
+need to import the p-net source project. The files ``.project`` and ? have been
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Done

File: doc/getting_started_rtkernel.rst [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+
+   help
+
+To start the sample application on target, type this command in the on-target
+shell::
+
+   pnio_run
+
+
+Adjust log level
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Moved to Linux page (which appear first). Added a reference to the Linux page here.

File: doc/linuxtiming.rst [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+For more details, see:
+
+* https://wiki.linuxfoundation.org/realtime/start
+* https://rt.wiki.kernel.org/index.php/Main_Page
+
+It is important to write you application so that it can use the benefits of
+the real-time patches. This includes running in a separate thread and setting
+the priorities properly.
+
+For the real-time patches to have an effect on p-net, set the ``USE_SCHED_FIFO``
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/linuxtiming.rst]: Done

File: doc/linuxtiming.rst [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+processor core. This assumes that you have more than one core in your CPU.
+Check it using::
+
+   cat /proc/cpuinfo
+
+By setting the ``isolcpus=1`` Linux boot command, the kernel will not put any
+processes on CPU core number 1 automatically. This option is typically set from
+the boot loader.
+
+To see which kernel command line options that have been used::
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/linuxtiming.rst]: Done

File: doc/getting_started_rtkernel.rst [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+Run on target
+-------------
+First start Segger J-link GDB debug server::
+
+    JLinkGDBServerExe
+
+Select "USB" and target device "XMC4800-2048" in the GUI. Use "Little Endian",
+Target interface "SWD" and speed "Auto Selection".
+
+Adjust debugger settings by right-clicking the Profinet build project,
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Modified the text.

File: doc/getting_started_rtkernel.rst [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+Run the compiled code on target by right-clicking the Profinet build project,
+and selecting "Debug as > Hardware debugging". Select ``pn_dev.elf`` and
+click OK. The download progress pop-up window should appear.
+
+To be able to view register content, use the MMR tab in the debug view. Select
+core "XMC4800".
+
+Open a terminal to view debug output from the target::
+
+    sudo picocom -b 115200 /dev/ttyACM0
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Modified text to use a terminal (but kept picocom as an example)

File: doc/prepare_raspberrypi.rst [hunk 26]
```diff
@@ -0,0 +1,147 @@
+Install Raspbian on the Raspberry Pi (by using a Linux laptop)
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/prepare_raspberrypi.rst]: Replaced parts of it with link to official documentation. Kept part on changing hostname, and permanently setting static IP address.

File: include/pnet_api.h (hunk 1 above)
Comment (pyhys) [reply include/pnet_api.h]: Modified so that it not will confuse Doxygen.

File: README.md (hunk 2 above)
Comment (pyhys) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (pyhys) [reply README.md]: Done

File: README.md (hunk 4 above)
Comment (pyhys) [reply README.md]: Changed to "handle full Ethernet frames"

File: README.md (hunk 5 above)
Comment (pyhys) [reply README.md]: Done

File: doc/getting_started_rtkernel.rst (hunk 6 above)
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Removed

File: README.md [hunk 27]
```diff
@@ -30,97 +41,38 @@ Features:
...
-=========
+* cmake 3.13 or later
 
- * Workbench 2017.1 or later
+For Linux:
 
-Set the following environment variables. You should use a bash shell,
-such as for instance the Command Line in your Toolbox installation.
+* gcc 4.6 or later
+* See the "Real-time properties of Linux" page in the documentation on how to
```
Comment (hefloryd) [inline README.md]: how to improve Linux timing

File: doc/getting_started_rtkernel.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (hefloryd) [reply doc/getting_started_rtkernel.rst]: This still seems incorrect. This should be enough:

Right-click on project
Choose Debug as > Hardware Debugging
Select JLink 
Add "monitor reset 0"
"Apply and close" or start debugging

File: doc/getting_started_rtkernel.rst [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+
+Run the compiled code on target by right-clicking the Profinet build project,
+and selecting "Debug as > Hardware debugging". Select ``pn_dev.elf`` and
+click OK. The download progress pop-up window should appear.
+
+To be able to view register content, use the MMR tab in the debug view. Select
+core "XMC4800".
+
+Open a terminal to view debug output from the target, which will appear as
+``/dev/ttyACM0``. For example picocom can be used (add yourself to the
```
Comment (hefloryd) [inline doc/getting_started_rtkernel.rst]: It may not be ttyACM0, if you have other serial devices attached.

Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Updated

File: README.md (hunk 27 above)
Comment (pyhys) [reply README.md]: Done

File: doc/compliancetest.rst [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+together.
+
+Each manufacturer of Profinet equipment needs a Vendor ID, which is obtained
+from the PI Certification Office.
+
+The tests checks the GSD file, the hardware interface and the Profinet
+behavior. The user manual must be available.
+The tests should be performed on a device from series production.
+
+The Profinet equipment also need to fulfill parts of the IEEE802 standards,
```
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/compliancetest.rst]: Done

File: doc/getting_started_rtkernel.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (pyhys) [reply doc/getting_started_rtkernel.rst]: Done

**Pull request merged**